### DISCUSSION TITLE
Detecting a partitioned run in a custom IO Manager with Single Run backfills

### DISCUSSION THREAD

**Discussion by cody-scott** [dagster-io/dagster#19098] [Q&A]
Hi, can anyone provide some feedback on the proper way to detect a partitioned run in a backfill, using the Single Run backfill? 

My IO manager is trying to check for a partition to build the clean up query. On a normal run it works fine, but when we run a backfill the `has_partition_key` is False. Asset uses `BackfillPolicy.single_run()` and a daily partition.

The dagster IO managers use a `TableSlice` class, but if i recall this isn't a public class.

```python
def get_cleanup_statement(table: str, schema: str, context: OutputContext) -> str:
    """Returns a SQL statement that deletes data in the given table to make way for the output data
    being written.
    """
    exist_block = f"SELECT * FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_SCHEMA = '{schema}' AND TABLE_NAME = '{table}'"
    if context.has_partition_key:
        part_txt = _partition_where_clause(context)
        delete_block = f"DELETE FROM {schema}.{table} {part_txt}".strip()
    else:
        delete_block = f"DELETE FROM {schema}.{table}".strip()

    return f"IF EXISTS ({exist_block}) {delete_block}"
```

Is there a better/preferred way of detecting partitioned runs?

Obviously in the current method, a backfill is going to drop the table which isn't ideal...

Thanks!

Using version 1.5.12.

**Comment by cody-scott** [comment]
Follow up, should i be using `has_asset_partitions` instead of `has_partition_key`?

**Reply by bolinzzz** [reply]
Our team is using `has_asset_partitions`, and it is working well for us.

You can actually launch a non partition asset with partition key via the launchpad, which will lead to `context.has_partition_key` to be true (whether you have provided a partition key, instead whether this asset is partitioned).
